pkgs/by-name/op/opencl-clang/package.nix

```diff
@@ -2,58 +2,74 @@
   lib,
   stdenv,
   applyPatches,
+  callPackage,
+  buildPackages,
+  targetPackages,
   fetchFromGitHub,
   cmake,
   git,
-  llvmPackages_15,
   spirv-llvm-translator,
-  buildWithPatches ? true,
 }:
 
 let
+  llvmPkgs =
+    let
+      llvmPkgs = callPackage ./llvm {
+        buildLlvmTools = buildPackages.opencl-clang.llvmPkgs.tools;
+        targetLlvmLibraries = targetPackages.opencl-clang.llvmPkgs.libraries or llvmPkgs.libraries;
+        targetLlvm = targetPackages.opencl-clang.llvmPkgs.llvm or llvmPkgs.llvm;
+        version = "15.0.7";
+        sha256 = "sha256-wjuZQyXQ/jsmvy6y1aksCcEDXGBjuhpgngF3XQJ/T4s=";
+      };
+    in
+    llvmPkgs;
+
+  patchesOut = stdenv.mkDerivation {
+    pname = "opencl-clang-patches";
+    inherit version src;
+    # Clang patches assume the root is the llvm root dir
+    # but clang root in nixpkgs is the clang sub-directory
+    postPatch = ''
+      for filename in patches/clang/*.patch; do
+        substituteInPlace "$filename" \
+          --replace-fail "a/clang/" "a/" \
+          --replace-fail "b/clang/" "b/"
+      done
+    '';
+
+    installPhase = ''
+      cp -r patches/ $out
+      mkdir -p $out/clang $out/llvm
+    '';
+  };
+
   addPatches =
     component: pkg:
     pkg.overrideAttrs (oldAttrs: {
       postPatch = oldAttrs.postPatch or "" + ''
-        for p in ${passthru.patchesOut}/${component}/*; do
+        for p in ${patchesOut}/${component}/*; do
           patch -p1 -i "$p"
         done
       '';
     });
 
-  llvmPkgs = llvmPackages_15;
-  inherit (llvmPkgs) llvm;
+  llvm = addPatches "llvm" llvmPkgs.tools.libllvm;
   spirv-llvm-translator' = spirv-llvm-translator.override { inherit llvm; };
-  libclang = if buildWithPatches then passthru.libclang else llvmPkgs.libclang;
+  libclang = addPatches "clang" llvmPkgs.libclang;
 
   passthru = rec {
+    inherit
+      llvmPkgs
+      llvm
+      libclang
+      patchesOut
+      ;
```

Comment:
Since you said `opencl-clang` is only used for `intel-graphics-compiler`, maybe it should be inlined into that package, too? I think it’s best to avoid exposing this single‐purpose LLVM build accessibly outside of `passthru.tests`, and the two derivations seem quite coupled, from what you’ve said.

But as I said in the initial comment, it’s likely you don’t even need a separate package here…